### ISSUE TITLE
Question: running Lemmy both on the Internet and in the Tor Network

### ISSUE THREAD

**Issue by gribodyr** [LemmyNet/lemmy#4140]
### Question

Hi guys, 

So I know it is possible to run Lemmy as a Tor hidden service, and of course as a regular web application.

However, what I have found is that you cannot do both at the same time. I mean, technically you can point the hidden service to the same Nginx instance, but then security issues start to creep in. I am going to list a few that I either found or thought of, but in no way is the list below exhaustive. 

I am not very well versed with Lemmy, so it took a bit of figuring out what's going on under the hood, including the default debug build setting being to strip debug symbols from the executable, lol. 

1. For uploaded images, Lemmy exposes absolute URLs to the client (https://github.com/LemmyNet/lemmy-ui/issues/2204). I figured that when you upload an image, its absolute URL is stored in the post or comment in the database. This leaks the underlying public domain name and/or IP address to the client.

2. If a user posts a link to an external image, it can be used to harvest end users' IP addresses. Yes, I know they will be protected by Tor, but still users might not realize that not all content on a Lemmy instance is hosted within .onion.

3. I need to double check, but lemmy-ui might refer to some other assets via a fully qualified domain name.
 
My use case might be sorta unique, but ultimately I would like to be able to run the same instance both on clearnet and in Tor. My user must be be sure that no content is hosted outside, and for the one that is (like an external link), they have full control _before_ the browser tries to load it. 

I've given this some thought, and a part of the solution that comes to mind is to implement an API proxy that would basically do  a few things:
- It would rewrite all absolute URLs to relative ones on the requests and responses, and for the Tor-sourced requests would ensure that the API responses reference only the .onion domain. 
- It would use the Tor network to download externally hosted media (we don't want to expose the server IP addresses to 3rd parties either), store it locally, and rewrite the original URLs with local ones. Think "transparent caching image proxy". 

Ultimately, this should be a Telegram-level security system with which it is practically impossible to upload any content capable of tracking both user activity and/or the server-side one. And I believe it should support a drop-in model where Lemmy itself can be updated with no effort, so that's why I am not considering a hardened fork for this. 

Also, I don't really believe we can have federation working within Tor. It kinda defeats the purpose of keeping stuff hidden. So this applies probably to a single-instance federation-off scenario.

I know this probably is a broad topic, but I would like to hear what the community thoughts on this are. 

Thank you!

**Comment by Nutomic** [comment]
> It would rewrite all absolute URLs to relative ones on the requests and responses, and for the Tor-sourced requests would ensure that the API responses reference only the .onion domain.

There is https://github.com/LemmyNet/lemmy/pull/4035 which rewrites all images to go through a local proxy. It currently uses absolute urls but that could be changed. Might be complicated though to get the correct domain for federated urls.

> It would use the Tor network to download externally hosted media (we don't want to expose the server IP addresses to 3rd parties either), store it locally, and rewrite the original URLs with local ones. Think "transparent caching image proxy".

Also the above, plus we recently made HTTP(S)_PROXY env vars working so you can route all outgoing traffic through Tor https://github.com/LemmyNet/lemmy/issues/3489.

> Also, I don't really believe we can have federation working within Tor. It kinda defeats the purpose of keeping stuff hidden. So this applies probably to a single-instance federation-off scenario.

Federation is just transferring JSON over HTTP so there is no fundamental problem doing it over Tor. Only problem is that Lemmy requires HTTPS for federation in production mode but that doesnt make sense in case of Tor.

**Comment by gribodyr** [comment]
> There is https://github.com/LemmyNet/lemmy/pull/4035 which rewrites all images to go through a local proxy. 

Yeah, I actually stumbled upon that after I made this post. Great stuff that you are already working on this! My first idea was to have an image proxy, but then I saw the absolute paths. 

> It currently uses absolute urls but that could be changed. Might be complicated though to get the correct domain for federated urls.

This might be a problem though. This prevents users from being able to access the same instance both from Tor and from clearnet. 

> Also check out https://github.com/salif/lemmy-onion-service

I guess I am going to have another fun session figuring out how to make this work in my LXC container :D 

J/k aside, I didn't have issues running through Tor when it is the only way to access the instance. I will need to double check though as I haven't yet tested running in Tor after figuring out my recent pict-rs/FUSE challenges. 

Thank you @Nutomic. I will report my findings here later.

**Comment by Nutomic** [comment]
> > It currently uses absolute urls but that could be changed. Might be complicated though to get the correct domain for federated urls.
>
> This might be a problem though. This prevents users from being able to access the same instance both from Tor and from clearnet.

That PR is not finished, its still possible to change this. For now Ive added a note about it.

> And my comment about federation was more about exposing [user@host.onion](mailto:user@host.onion). It might be non-obvious at least to my users that their content could get federated somewhere else.

Im not entirely sure what you mean by this. If you mean federation over tor, thats not such a good idea because most instances wont be able to connect to onion addresses. Better to federate over clearnet unless you have very high security requirements.
